File: go/internal/agent/summary/summary_test.go
package summary

import (
	"context"
	"errors"
	"iter"
	"os"
	"strings"
	"testing"
	"time"

	"google.golang.org/adk/model"

	"github.com/jkjamies/automation-agent/internal/agent/setup"
	"github.com/jkjamies/automation-agent/internal/githubapi"
	"github.com/jkjamies/automation-agent/internal/notify"
)

// --- fakes ---

type fakeLister struct {
	byRepo map[string][]githubapi.Commit
	err    error
}

func (f fakeLister) ListCommitsSince(_ context.Context, owner, repo string, _ time.Time) ([]githubapi.Commit, error) {
	if f.err != nil {
		return nil, f.err
	}
	return f.byRepo[owner+"/"+repo], nil
}

type fakeNotifier struct{ msgs []notify.Message }

func (n *fakeNotifier) Notify(_ context.Context, m notify.Message) error {
	n.msgs = append(n.msgs, m)
	return nil
}

// stubLLM satisfies model.LLM without producing output (structure tests don't run
// the agent). It deliberately avoids importing genai to honor the ARCH rule.
type stubLLM struct{}

func (stubLLM) Name() string { return "stub" }
func (stubLLM) GenerateContent(context.Context, *model.LLMRequest, bool) iter.Seq2[*model.LLMResponse, error] {
	return func(func(*model.LLMResponse, error) bool) {}
}

type fakeState map[string]any

func (f fakeState) All() iter.Seq2[string, any] {
	return func(yield func(string, any) bool) {
		for k, v := range f {
			if !yield(k, v) {
				return
			}
		}
	}
}

// --- tests ---

func TestFormatCommits(t *testing.T) {
	if got := formatCommits("o/r", nil); !strings.Contains(got, "no commits") {
		t.Errorf("empty = %q", got)
	}
	commits := []githubapi.Commit{{SHA: "abcdef1234", Message: "fix bug\n\ndetails", Author: "Jane"}}
	got := formatCommits("o/r", commits)
	if !strings.Contains(got, "abcdef1") || !strings.Contains(got, "fix bug") || !strings.Contains(got, "Jane") {
		t.Errorf("got = %q", got)
	}
	if strings.Contains(got, "details") {
		t.Errorf("should keep only the first line: %q", got)
	}
}

func TestBuildInstruction(t *testing.T) {
	st := fakeState{
		"commits:b/b": "repo B data",
		"commits:a/a": "repo A data",
		"other":       "ignore me",
	}
	got := buildInstruction("PROMPT", st)
	if !strings.HasPrefix(got, "PROMPT") {
		t.Errorf("prompt should lead: %q", got)
	}
	if strings.Contains(got, "ignore me") {
		t.Error("non-commits state key leaked into instruction")
	}
	if ai, bi := strings.Index(got, "repo A data"), strings.Index(got, "repo B data"); ai < 0 || bi < 0 || ai > bi {
		t.Errorf("commit sections not sorted: a=%d b=%d", ai, bi)
	}
}

func TestBuildInstructionEmpty(t *testing.T) {
	if got := buildInstruction("P", fakeState{}); !strings.Contains(got, "no commit data") {
		t.Errorf("got = %q", got)
	}
}

func TestSplitRepoAndSafeName(t *testing.T) {
	if _, _, ok := splitRepo("owner/repo"); !ok {
		t.Error("valid repo rejected")
	}
	if _, _, ok := splitRepo("bad"); ok {
		t.Error("invalid repo accepted")
	}
	if got := setup.SafeName("a/b:c"); got != "a_b_c" {
		t.Errorf("SafeName = %q, want a_b_c", got)
	}
}

func TestBuildSummaryAgentStructure(t *testing.T) {
	a, err := BuildSummaryAgent(Deps{
		LLM: stubLLM{}, GH: fakeLister{}, Notify: &fakeNotifier{}, Repos: []string{"o/r"},
	})
	if err != nil {
		t.Fatalf("build: %v", err)
	}
	if a.Name() != "summary_workflow" {
		t.Errorf("name = %q, want summary_workflow", a.Name())
	}
}

func TestBuildSummaryAgentValidation(t *testing.T) {
	if _, err := BuildSummaryAgent(Deps{LLM: stubLLM{}, GH: fakeLister{}, Notify: &fakeNotifier{}}); err == nil {
		t.Error("no repos should error")
	}
	if _, err := BuildSummaryAgent(Deps{Repos: []string{"o/r"}}); err == nil {
		t.Error("nil deps should error")
	}
}

// TestWorkflowRunWithStub drives the whole workflow through a real runner with a
// stub LLM (no live model), exercising the fetch and notify code agents and the
// runner plumbing deterministically.
func TestWorkflowRunWithStub(t *testing.T) {
	gh := fakeLister{byRepo: map[string][]githubapi.Commit{
		"o/r": {{SHA: "abc1234", Message: "do the thing", Author: "X"}},
	}}
	notifier := &fakeNotifier{}
	a, err := BuildSummaryAgent(Deps{LLM: stubLLM{}, GH: gh, Notify: notifier, Repos: []string{"o/r"}})
	if err != nil {
		t.Fatalf("build: %v", err)
	}
	r, err := setup.NewRunner("stub-test", a)
	if err != nil {
		t.Fatalf("NewRunner: %v", err)
	}
	if err := setup.Drive(context.Background(), r, "u", "s", "go"); err != nil {
		t.Fatalf("Drive: %v", err)
	}
	// The stub produces no digest, so the notifier posts the fallback — but it must
	// still have been invoked exactly once.
	if len(notifier.msgs) != 1 {
		t.Fatalf("notifications = %d, want 1", len(notifier.msgs))
	}
}

func TestWorkflowFetchError(t *testing.T) {
	gh := fakeLister{err: errors.New("api down")}
	a, _ := BuildSummaryAgent(Deps{LLM: stubLLM{}, GH: gh, Notify: &fakeNotifier{}, Repos: []string{"o/r"}})
	r, _ := setup.NewRunner("stub-test", a)
	if err := setup.Drive(context.Background(), r, "u", "s", "go"); err == nil {
		t.Fatal("expected an error from a failing fetch")
	}
}

// TestLiveSummaryWorkflow runs the whole workflow end-to-end against a real Ollama
// server (opt-in via OLLAMA_LIVE), validating fetch -> state -> summarizer(LLM) ->
// OutputKey -> notify. Asserts only that a non-empty digest is posted.
func TestLiveSummaryWorkflow(t *testing.T) {
	if os.Getenv("OLLAMA_LIVE") == "" {
		t.Skip("set OLLAMA_LIVE=1 to run the live summary workflow")
	}
	tag := os.Getenv("OLLAMA_MODEL")
	if tag == "" {
		tag = "gemma4:e4b"
	}
	llm, err := setup.NewOllamaModel("http://localhost:11434", tag)
	if err != nil {
		t.Fatal(err)
	}

	gh := fakeLister{byRepo: map[string][]githubapi.Commit{
		"acme/api": {{SHA: "1111111", Message: "add request rate limiting", Author: "Jane"}},
		"acme/web": {{SHA: "2222222", Message: "fix login redirect loop", Author: "Ravi"}},
	}}
	notifier := &fakeNotifier{}

	a, err := BuildSummaryAgent(Deps{LLM: llm, GH: gh, Notify: notifier, Repos: []string{"acme/api", "acme/web"}})
	if err != nil {
		t.Fatal(err)
	}

	r, err := setup.NewRunner("summary-test", a)
	if err != nil {
		t.Fatal(err)
	}
	if err := setup.Drive(context.Background(), r, "u", "s", "Run the daily digest."); err != nil {
		t.Fatalf("drive: %v", err)
	}

	if len(notifier.msgs) != 1 {
		t.Fatalf("expected 1 notification, got %d", len(notifier.msgs))
	}
	if text := strings.TrimSpace(notifier.msgs[0].Text); text == "" || strings.Contains(text, "no digest") {
		t.Fatalf("empty or failed digest: %q", text)
	}
	t.Logf("digest:\n%s", notifier.msgs[0].Text)
}

File: kotlin/src/test/kotlin/io/github/jkjamies/automationagent/agent/summary/SummaryTest.kt
package io.github.jkjamies.automationagent.agent.summary

import com.google.adk.kt.models.LlmRequest
import com.google.adk.kt.models.LlmResponse
import com.google.adk.kt.models.Model
import io.github.jkjamies.automationagent.agent.setup.drive
import io.github.jkjamies.automationagent.agent.setup.newRunner
import io.github.jkjamies.automationagent.agent.setup.safeName
import io.github.jkjamies.automationagent.githubapi.Commit
import io.github.jkjamies.automationagent.notify.Message
import io.github.jkjamies.automationagent.notify.Notifier
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.matchers.collections.shouldHaveSize
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotContain
import io.kotest.matchers.string.shouldStartWith
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emptyFlow
import java.time.Instant

private class FakeLister(
    private val byRepo: Map<String, List<Commit>> = emptyMap(),
    private val error: Throwable? = null,
) : CommitLister {
    override suspend fun listCommitsSince(owner: String, repo: String, since: Instant): List<Commit> {
        error?.let { throw it }
        return byRepo["$owner/$repo"] ?: emptyList()
    }
}

private class FakeNotifier : Notifier {
    val msgs = mutableListOf<Message>()
    override suspend fun notify(message: Message) {
        msgs += message
    }
}

/** A Model that produces no output — structure/plumbing tests never need real generation. */
private object StubModel : Model {
    override val name: String = "stub"
    override fun generateContent(request: LlmRequest, stream: Boolean): Flow<LlmResponse> = emptyFlow()
}

private fun commit(sha: String, message: String, author: String): Commit =
    Commit(sha = sha, message = message, author = author, url = "", at = Instant.EPOCH)

class SummaryTest : BehaviorSpec({
    Given("the formatCommits helper") {
        When("formatting empty and populated commit lists") {
            Then("it summarizes one line per commit and keeps only the first message line") {
                formatCommits("o/r", emptyList()) shouldContain "no commits"
                val got = formatCommits("o/r", listOf(commit("abcdef1234", "fix bug\n\ndetails", "Jane")))
                got shouldContain "abcdef1"
                got shouldContain "fix bug"
                got shouldContain "Jane"
                got shouldNotContain "details"
            }
        }
    }

    Given("commit data in state") {
        When("building the instruction") {
            Then("the prompt leads, non-commit keys are ignored, and sections are sorted") {
                val state = mapOf<String, Any?>("commits:b/b" to "repo B data", "commits:a/a" to "repo A data", "other" to "ignore me")
                val got = buildInstruction("PROMPT", state)
                got shouldStartWith "PROMPT"
                got shouldNotContain "ignore me"
                got.indexOf("repo A data").let { a -> got.indexOf("repo B data").let { b -> (a in 0 until b) shouldBe true } }
            }
        }
        When("there is no commit data") {
            Then("it says so") {
                buildInstruction("P", emptyMap()) shouldContain "no commit data"
            }
        }
    }

    Given("repo strings") {
        When("splitting and sanitizing") {
            Then("valid repos split, invalid ones reject, and names are sanitized") {
                splitRepo("owner/repo") shouldBe ("owner" to "repo")
                splitRepo("bad") shouldBe null
                safeName("a/b:c") shouldBe "a_b_c"
            }
        }
    }

    Given("valid summary deps") {
        When("building the workflow") {
            Then("the composed agent is named summary_workflow") {
                val a = buildSummaryAgent(SummaryDeps(StubModel, FakeLister(), FakeNotifier(), listOf("o/r")))
                a.name shouldBe "summary_workflow"
            }
        }
    }

    Given("summary deps with no repos") {
        When("building the workflow") {
            Then("it fails fast") {
                shouldThrow<IllegalArgumentException> {
                    buildSummaryAgent(SummaryDeps(StubModel, FakeLister(), FakeNotifier(), emptyList()))
                }
            }
        }
    }

    Given("a workflow driven through a real runner with a stub LLM") {
        When("running it") {
            Then("the notifier is invoked exactly once (the fallback digest)") {
                val gh = FakeLister(mapOf("o/r" to listOf(commit("abc1234", "do the thing", "X"))))
                val notifier = FakeNotifier()
                val a = buildSummaryAgent(SummaryDeps(StubModel, gh, notifier, listOf("o/r")))
                drive(newRunner("stub-test", a), "u", "s", "go")
                notifier.msgs shouldHaveSize 1
            }
        }
    }

    Given("a workflow whose fetch fails") {
        When("running it") {
            Then("the error propagates") {
                val a = buildSummaryAgent(SummaryDeps(StubModel, FakeLister(error = RuntimeException("api down")), FakeNotifier(), listOf("o/r")))
                shouldThrow<RuntimeException> { drive(newRunner("stub-test", a), "u", "s", "go") }
            }
        }
    }
})
